Unclear is offsetting one element past array length is allowed

### Location

https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

### Summary

[Offset](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset) doc states that 

> If the computed offset is non-zero, then self must be [derived from](https://doc.rust-lang.org/std/ptr/index.html#provenance) a pointer to some [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object), and the entire memory range between self and the result must be in bounds of that allocated object.

allocated object doc states that

> For all addresses a in addresses, a is in the range base .. (base + size) (note that this requires a < base + size, not a <= base + size)

so as i understand offsetting pointer to the head of an array one past it's length should not be allowed, but in practice it is allowed

```rust
fn main() {
    static A: [u8; 8] = [1, 2, 3, 4, 5, 6, 7, 8];
    static mut B: *const u8 = unsafe {A.as_ptr().offset(8)};
}
```
this code compiles just fine, but this one produces an error

```rust
fn main() {
    static A: [u8; 8] = [1, 2, 3, 4, 5, 6, 7, 8];
    static mut B: *const u8 = unsafe {A.as_ptr().offset(9)};
}
```

```console
error[E0080]: could not evaluate static initializer
   --> /home/daria/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:455:18
    |
455 |         unsafe { intrinsics::offset(self, count) }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ out-of-bounds pointer arithmetic: expected a pointer to 9 bytes of memory, but got alloc1 which is only 8 bytes from the end of the allocation
    |
note: inside `std::ptr::const_ptr::<impl *const u8>::offset`
   --> /home/daria/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:455:18
    |
455 |         unsafe { intrinsics::offset(self, count) }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `B`
   --> src/main.rs:3:39
    |
3   |     static mut B: *const u8 = unsafe {A.as_ptr().offset(9)};
    |                                       ^^^^^^^^^^^^^^^^^^^^

```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"xizheyin"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/library/core/src/ptr/const_ptr.rs
+++ b/library/core/src/ptr/const_ptr.rs
@@ -386,7 +386,8 @@ impl<T: ?Sized> *const T {
     /// * If the computed offset is non-zero, then `self` must be [derived from][crate::ptr#provenance] a pointer to some
     ///   [allocated object], and the entire memory range between `self` and the result must be in
     ///   bounds of that allocated object. In particular, this range must not "wrap around" the edge
-    ///   of the address space.
+    ///   of the address space. Note that "range" here refers to a half-open range as usual in Rust,
+    ///   i.e., `self..result` for non-negative offsets and `result..self` for negative offsets.
     ///
     /// Allocated objects can never be larger than `isize::MAX` bytes, so if the computed offset
     /// stays in bounds of the allocated object, it is guaranteed to satisfy the first requirement.

